### PR TITLE
[BEAM-773] Implement Metrics support for Flink runner

### PR DESCRIPTION
**Pull request by JingsongLi**
@iemejia Hope useful to you.

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by iemejia** [comment]
Thanks @JingsongLi !

I have not yet started with this but I plan to do it once I finish my pending tasks. Were you actively working on metrics at this moment ? Because if this is the case you can take the ticket. I just took it because I thought nobody was working on it and I want to make it progress. WDYT ?

In case you don't have the time to finish this at this moment, I will rebase from your branch but I will probably take more time to finish it than you (considering your current progress).

**Comment by JingsongLi** [comment]
@iemejia Thank you for the reply, I will take the time to finish it. I reassign the jira to me.

**Comment by iemejia** [comment]
Excellent, there is a big amount of interest on having this one, not only for the improvements but to unblock the removal of Aggregators (https://issues.apache.org/jira/browse/BEAM-775). I am eager to have it too to move my work on benchmarks with Nexmark to use metrics.
I will follow the issue and don't hesitate to ping me if you need a hand with something.

**Comment by JingsongLi** [comment]
I like benchmarks. This is very useful for the performance optimization of sdk and runners. I'm so glad you can give me some suggestions too.

**Comment by iemejia** [comment]
This issue and the one on the interval windows I reported yesterday are part of our ongoing work on recovering Nexmark and getting it to work with all the runners, in case you want to take a look it is here:
https://github.com/iemejia/beam/tree/BEAM-160-nexmark/integration/java/nexmark

It is still a WIP and we will be updating the support matrix as we are chasing the fixes in the different runners (and running our executions).

**Comment by aljoscha** [comment]
Thanks a lot for working on this @JingsongLi 😃 

I'm currently traveling (on vacation) so I'll be rather slow with reviewing, unfortunately.

**Comment by aljoscha** [comment]
A quick early remark: Could we add a configuration value to `FlinkPipelineOptions` that allows completely deactivating metrics? I think it should be easily doable because we can just not wrap in the decorating `DoFnRunner`.

**Comment by iemejia** [comment]
Big +1 for the disable metrics idea, I had already created a JIRA to extend this to the other runners, so Flink would be the first.
https://issues.apache.org/jira/browse/BEAM-1758

**Comment by dhalperi** [comment]
R: @aljoscha -- I assume you're handling pushing this through?

**Comment by aljoscha** [comment]
@dhalperi Yes, I'm a bit slow, though, since I'm currently on vacation.

**Comment by JingsongLi** [comment]
@aljoscha There is a exception when run the {{MetricsTest.testBoundedSourceMetrics}}.
Caused by: java.lang.IllegalStateException: No operators defined in streaming topology. Cannot execute.
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.getStreamGraph(StreamExecutionEnvironment.java:1515)
	at org.apache.flink.streaming.api.environment.LocalStreamEnvironment.execute(LocalStreamEnvironment.java:86)
	at org.apache.beam.runners.flink.FlinkPipelineExecutionEnvironment.executePipeline(FlinkPipelineExecutionEnvironment.java:113)
	at org.apache.beam.runners.flink.FlinkRunner.run(FlinkRunner.java:116)
	... 29 more

Should we add an empty ParDo when pipeline just have only one Source.

**Comment by davorbonaci** [comment]
@aljoscha, any comments perhaps?

**Comment by aljoscha** [comment]
Now getting back to reviews after busy weeks with Flink Forward.

**Comment by aljoscha** [comment]
@JingsongLi I finally had time to look at this. I think it's very good! 👍 

The addition of IO Metrics, is this covered by a Beam Jira Issue?

**Comment by JingsongLi** [comment]
There is no separate issue of Flink runner, the related issue is: 
https://issues.apache.org/jira/browse/BEAM-1397

**Comment by JingsongLi** [comment]
@aljoscha Welcome back~

**Comment by aljoscha** [comment]
Thanks! 😄 And ok, I see.

**Comment by aljoscha** [comment]
@JingsongLi 

I rebased your work on top of the (very) recent Splittable DoFn and Flink module reorganisation changes: https://github.com/aljoscha/beam/commits/finish-pr-2300-flink-metrics

I also tried a different approach for the source metrics. I didn't like all the `getDelegate()` methods to much so I changed it into an invocation util. I still don't like that we have special invoke calls, though. What do you think?

**Comment by JingsongLi** [comment]
Very nice. That is better. This will not affect the call of other methods.

**Comment by aljoscha** [comment]
There is still the problem of the failing test if there's only a source. Have you looked into how hard it is to generate a dummy operator after the source if we only have a source?

**Comment by JingsongLi** [comment]
I add checkGenerateDummyOperator method in `FlinkStreamingPipelineTranslator`.

**Comment by aljoscha** [comment]
I'm afraid that only works for the one simple test case. What happens if there are two sources without any other operators? I think we'll then get the same exception.

I think we need to use the Pipeline surgery API to find out which sources don't have any successor operations and then create dummy operations for those.

**Comment by JingsongLi** [comment]
Yes, and Flink Streaming mode just execute the source which has child nodes. So we must find out all lonely sources.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java [hunk 1]
```diff
@@ -112,10 +115,61 @@ public void translate(Pipeline pipeline) {
...
+          }
+        }
+      }
+    });
+
+    // add dummy parDo
+    for (Map.Entry<PValue, Boolean> entry : sources.entrySet()) {
+      if (!entry.getValue()) {
+        PCollection<?> pCollection = (PCollection) entry.getKey();
+        pCollection.apply(ParDo.of(new EmptyDoFn<>()));
```
Comment (aljoscha) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: @JingsongLi This seems to do the trick. 👍 I just want to double check with @tgroh  if there is a obvious canonical way of doing this that we're missing.

@tgroh Do you have any opinion on this, as the master Pipeline surgeon? 😄

Comment (tgroh) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Turns out I did basically exactly this yesterday in https://github.com/apache/beam/pull/2587

Comment (tgroh) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: Ultimately it's `UnconsumedReads.ensureAllReadsConsumed(pipeline)`

Comment (JingsongLi) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: That's great!

**Comment by aljoscha** [comment]
Run Flink ValidatesRunner

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java (hunk 1 above)
Comment (aljoscha) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java]: I knew I was asking the right person... 😃

**Comment by aljoscha** [comment]
Thanks a lot @JingsongLi for making sure we finally have Metrics support! ~3q

**Comment by iemejia** [comment]
Thanks a lot @JingsongLi for this one, perfect timing!

**Pull request closed without merging**